dlib/image_processing/box_overlap_testing.h

```diff
@@ -17,11 +17,11 @@ namespace dlib
         const drectangle& b
     ) 
     {
-        const double inner = a.intersect(b).area();
+        const auto inner = a.intersect(b).area();
         if (inner == 0)
             return 0;
-        const double outer = (a+b).area();
-        return inner/outer;
+        const double outer = double(a.area() + b.area() - inner);
+        return double(inner)/outer;
```

Comment:
Since we are passing `drectangle`s as parameters, we don't need those casts, or the explicit type for `outer`.

Reply:
Correct, saw my mistake a few minutes ago. Will change it asap.
Furthermore, the check if inner == 0 should be replaced as comparing doubles with 0 zero is risky.

Reply:
I think that the check with 0 is fine.

Reply:
The thing is that `rectangle` and `drectangle` are either empty or have an area larger to 1, as you can see here:

https://github.com/davisking/dlib/blob/eb5de0d534f637253a7b34b01048b655d5befc20/dlib/geometry/drectangle.h#L95-L120

So, calling the `drectangle::area()` method is always going to return either 0 exactly or a number larger than 1.

Reply:
There isn't anything funny or dangerous about comparing to zero.